### PR TITLE
Refactor to Bot interface

### PR DESCRIPTION
**Pull request by hbradio**
This refactors all of the bot-related peripherals (i.e. those not on the micro:bit itself) to be accessed through a new `Bot` interface. This made main.cpp much cleaner.

I've tested that everything still seems to work on the Gigglebot. I haven't tested the Cutebot stuff yet.

### CODE REVIEW

File: inc/drivers/Bot.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class Bot {
+
+  bool detected;
+
+  public:
+
+  virtual void setRightMotorPower(int power) = 0;
+  virtual void setLeftMotorPower(int power) = 0;
+  virtual void setBothMotorsPower(int power) = 0;
+  virtual void setLeftHeadlightColor(uint8_t red, uint8_t green, uint8_t blue);
```
Comment (cabarnes) [inline inc/drivers/Bot.h]: Shouldn't these be `=0` too?

File: inc/drivers/cutebot/CutebotMotor.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include "inc/drivers/Motor.h"
+
+enum CutebotMotorId {
+  CUTEBOT_MOTOR_ID_RIGHT = 1,
+  CUTEBOT_MOTOR_ID_LEFT,
+};
+
+class CutebotMotor : public Motor
+{
+    MicroBitI2C i2c;
```
Comment (cabarnes) [inline inc/drivers/cutebot/CutebotMotor.h]: Should this be `&i2c`?

File: source/main.cpp [hunk 3]
```diff
@@ -161,6 +155,14 @@ int main()
     // Initialise the micro:bit runtime.
     uBit.init();
 
+    Gigglebot* gigglebot = new Gigglebot(uBit.i2c);
+    Cutebot* cutebot = new Cutebot(uBit.i2c);
```
Comment (cabarnes) [inline source/main.cpp]: Might want to only construct the second one of the first isn't detected. Then delete the first so there isn't extra memory usage

File: inc/drivers/cutebot/CutebotMotor.h (hunk 2 above)
Comment (hbradio) [reply inc/drivers/cutebot/CutebotMotor.h]: I think you're right. I'm not sure how it compiled like that. (And worked for the GigglebotMotor?)

File: source/main.cpp (hunk 3 above)
Comment (hbradio) [reply source/main.cpp]: :+1:

Comment (hbradio) [reply source/main.cpp]: `warning: deleting object of polymorphic class type 'Gigglebot' which has non-virtual destructor might cause undefined behavior`

YOLO

Comment (cabarnes) [reply source/main.cpp]: I removed this warning by implementing the virtual destructor

**Review by cabarnes — APPROVED** [review]

**Pull request merged**